utils/s2n_result.h

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#pragma once
+
+#include <s2n.h>
+#include <stdbool.h>
+
+/* A value which indicates the outcome of a function */
+typedef struct _s2n_result s2n_result;
+
+/* used to signal a successful function return */
+extern const s2n_result S2N_RESULT_OK;
+/* used to signal an error while executing a function */
+extern const s2n_result S2N_RESULT_ERROR;
+
+/* returns true when the result is S2N_RESULT_OK */
+__attribute__((warn_unused_result)) bool s2n_result_is_ok(s2n_result result);
+
+/* returns true when the result is S2N_RESULT_ERROR */
+__attribute__((warn_unused_result)) bool s2n_result_is_error(s2n_result result);
+
+/* used in function declarations to signal function fallibility */
+#define S2N_RESULT __attribute__((warn_unused_result)) s2n_result
+
+/* s2n_result GUARD helpers */
+/* note: eventually this will just alias GUARD and be deprecated once everything is using s2n_result */
+#define GUARD_RESULT( x )               do {if ( s2n_result_is_error(x) ) return S2N_RESULT_ERROR;} while (0)
```

Comment:
Shouldn't it return `x`?

Reply:
I copied the pattern from the existing GUARD macro. Not sure it matters much.